transcom/mymove
# B 21492 INT2 - Part 2 - Improved Validation in Admin App for Transportation Office Assignments

## [B-21492](https://www13.v1host.com/USTRANSCOM38/assetdetail.v1?number=B-21492)

## Summary

Adds addition validation for B-21492 to prevent assigning two transportation offices in the same GBLOC.

## Verification Steps for the Author

These are to be checked by the author.


## Verification Steps for Reviewers

These are to be checked by a reviewer.


### Setup to Run the Code

- [Instructions for starting storybook](https://transcom.github.io/mymove-docs/docs/frontend/setup/storybook)
- [Instructions for starting the MilMove application](https://transcom.github.io/mymove-docs/docs/getting-started/application-setup/)
- [Instructions for running tests](https://transcom.github.io/mymove-docs/docs/getting-started/development/testing)

### How to test

1. Access the admin app.
2. Login as a new or existing admin user.
3. Open the form to create a new office user.
4. Attempt to create an office user with duplicate transportation office assignments or two transportation offices in the same GBLOC.
5. Save a valid office user.
6. Repeat step 4 with the edit form.
7. Try to break it. The validation messages don't refresh properly due to a quirk of a dependency of the version of react-admin (4.12, latest is 5.3) that we use. To avoid misleading error messages I have updated these forms to validate on submit rather than blur, which seems to have no issues with displaying the desired error message.

If you wish to verify the creation of your transportation office assignments in the DB the following query may be helpful.
`SELECT * FROM public.transportation_office_assignments toas
JOIN transportation_offices tos ON toas.transportation_office_id = tos.id
JOIN office_users ou ON ou.id = toas.id
WHERE ou.last_name = 'O''Testoghue'`

You could also test these changes on the B-21490 PR which should contain these changes and will allow you to switch between the assigned transportation offices.

### Frontend


### Backend


### Database

#### Any new migrations/schema changes:


## Screenshots
<img width="559" alt="Screenshot 2024-10-29 at 6 03 40 PM" src="https://github.com/user-attachments/assets/15c76f90-0979-4dea-8e26-29cdf9786af3">
